### PR TITLE
add auto-async execution to query pipelines

### PR DESCRIPTION
**Pull request by jerryjliu**
added a sample notebook here as well

### CODE REVIEW

**Comment by jerryjliu** [comment]
oops forgot to commit the actual code lol, just did

File: docs/examples/pipeline/query_pipeline_async.ipynb [hunk 1]
```diff
@@ -0,0 +1,306 @@
+{
```
Comment (logan-markewich) [inline docs/examples/pipeline/query_pipeline_async.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #26.        retrievers[str(chunk_size)] = vector_index.as_retriever()
</pre><p>I don't think the retrievers dict is used. Can we remove it?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/run-llama/llama_index/pull/9967/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (logan-markewich) [inline docs/examples/pipeline/query_pipeline_async.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #12.    # add links from input to retriever (id'ed by chunk_size)
</pre><p>nit: we aren't linking to a retriever in this case, but to a query engine? Did you mean to have it be retrievers?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/run-llama/llama_index/pull/9967/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

File: llama_index/query_pipeline/query.py [hunk 2]
```diff
@@ -426,18 +449,44 @@ async def _arun_multi(self, module_input_dict: Dict[str, Any]) -> Dict[str, Any]
...
-            )
+            # create tasks from popped nodes
+            tasks = []
+            for module_key in popped_nodes:
+                module = self.module_dict[module_key]
+                module_input = all_module_inputs[module_key]
+                tasks.append(module.arun_component(**module_input))
+
+            # run tasks
+            output_dicts = await asyncio.gather(*tasks)
```
Comment (logan-markewich) [inline llama_index/query_pipeline/query.py:483]: should we be using async `run_jobs` to avoid rate-limits?

**Review by logan-markewich — APPROVED** [review]
lgtm, just a few small comments

File: docs/examples/pipeline/query_pipeline_async.ipynb (hunk 1 above)
Comment (jerryjliu) [reply docs/examples/pipeline/query_pipeline_async.ipynb:1]: good catch. doing query engine since easier to measure latency speedup

**Pull request merged**